provisioner/terraform/tfparse/tfparse.go

```diff
@@ -247,28 +251,39 @@ func WriteArchive(bs []byte, mimetype string, path string) error {
 	return nil
 }
 
-// VariableDefaults returns the default values for all variables passed to it.
-func (p *Parser) VariableDefaults(ctx context.Context) (map[string]string, error) {
+// VariableDefaults returns the default values for all variables referenced in the values of tags.
+func (p *Parser) VariableDefaults(ctx context.Context, tags map[string]string) (map[string]string, error) {
+	var skipped []string
 	// iterate through vars to get the default values for all
-	// variables.
+	// required variables.
 	m := make(map[string]string)
 	for _, v := range p.module.Variables {
 		if v == nil {
 			continue
 		}
+		var found bool
+		for _, tv := range tags {
+			if strings.Contains(tv, v.Name) {
```

Comment:
won't a `strings.Contains` also contain variable names that are subsets?

Like `var.Foo` will match both `Foo` and `var.FooBar`

Reply:
Here is a test case:
```golang
			{
				name: "overlapping var name",
				files: map[string]string{
					`main.tf`: `
						// This file is the same as the above, except for this comment.
						variable "a" {
							type = string
							default = "1"
						}
						variable "ab" {
						 description = "This is a variable of type string"
						 type        = string
						 default     = jsonencode(["a", "b"])
						}
						data "coder_workspace_tags" "tags" {
							tags = {
								"foo": "bar",
								"a": var.a,
							}
						}`,
				},
				reqTags:  map[string]string{"foo": "noclobber"},
				wantTags: map[string]string{"owner": "", "scope": "organization", "foo": "bar", "a": "1"},
			},
```

Reply:
Gotcha, we may then need to do some more "intelligent" parsing.
I hesitate to reach for a regexp.... but `\b` may be exactly what we need here

Reply:
Is the reason you do a contains because it could be an expression like `foo(var.name)`?

Reply:
Is there any way to parse the expression and get the actual tokens? Or we only have `string` in the HCL?

Reply:
Yep. I'm looking into HCL traversals now as they might allow us to do this cleanly.

Reply:
OK, I've modified the implementation to instead:
- Gather the set of variables referenced by the tags up-front,
- Only evaluate the default value(s) for `coder_parameters` referenced by those tags.
- As default values of coder_parameters can reference other variables, to avoid needing to perform a graph traversal of all the dependencies here I've elected to simply continue passing in all the variable default values. We don't immediately enforce default values for all variables.
- To avoid unrelated variables with default values of complex types being rejected, I added some 'last-ditch' support for JSON-encoding them.